[systeminfo] Migrate to BND

falkena opened the pull request:
Signed-off-by: Alexander Falkenstern <alexander.falkenstern@gmail.com>

davidgraeff commented on bundles/org.openhab.binding.systeminfo/.classpath:
```diff
@@ -0,0 +1,33 @@
+			<attribute name="maven.pomderived" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="lib" path="lib/Moka7-1.0.2_io_patch.jar"/>
```
copy/paste error I guess. can be removed

falkena replied:
Ohh man. Thanks :-)

davidgraeff commented on bundles/org.openhab.binding.systeminfo/pom.xml:
```diff
@@ -0,0 +1,34 @@
+    <dependency>
+      <groupId>com.github.oshi</groupId>
+      <artifactId>oshi-core</artifactId>
+      <version>3.13.0</version>
```
add `<scope>provided</scope>` to all of them

falkena replied:
Done

J-N-K commented on features/karaf/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -847,7 +847,10 @@
         <feature>openhab-runtime-base</feature>
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.systeminfo/${project.version}</bundle>
+        <bundle dependency="true">wrap:mvn:com.github.oshi/oshi-core/3.13.0</bundle>
+        <bundle dependency="true">wrap:mvn:net.java.dev.jna/jna/5.2.0</bundle>
```
wrapping should not be necessary, this is already an OSGi bundle

falkena replied:
Played around to solve travis issues. Sadly, still no progress.

J-N-K commented on features/karaf/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -847,7 +847,10 @@
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.systeminfo/${project.version}</bundle>
+        <bundle dependency="true">wrap:mvn:com.github.oshi/oshi-core/3.13.0</bundle>
+        <bundle dependency="true">wrap:mvn:net.java.dev.jna/jna/5.2.0</bundle>
+        <bundle dependency="true">wrap:mvn:net.java.dev.jna/jna-platform/5.2.0</bundle>
```
wrapping should not be necessary, this is already an OSGi bundle

falkena replied:
Played around to solve travis issues. Sadly, still no progress.

J-N-K replied:
I don‘t think Travis failure is related to your changes. Travis fails on all PR.

falkena replied:
My local build fails too :-)

J-N-K commented on features/karaf/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -847,7 +847,10 @@
     <feature name="openhab-binding-systeminfo" description="System Info Binding" version="${project.version}">
         <feature>openhab-runtime-base</feature>
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.systeminfo/${project.version}</bundle>
+        <bundle dependency="true">wrap:mvn:com.github.oshi/oshi-core/3.13.0</bundle>
```
wrapping should not be necessary, this is already an OSGi bundle

falkena replied:
Played around to solve travis issues. Sadly, still no progress.

falkena commented:
I stuck with karaf feature verification: openHAB Add-ons :: Features :: Karaf :: Add-ons .... FAILURE
The message is 

> Feature resolution failed for [openhab-binding-systeminfo/2.5.0.SNAPSHOT]
> [ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-systeminfo; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-systeminfo)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-systeminfo/2.5.0.SNAPSHOT: missing requirement [openhab-binding-systeminfo/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.systeminfo; type=osgi.bundle; version="[2.5.0.201904152011,2.5.0.201904152011]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.systeminfo/2.5.0.201904152011: missing requirement [org.openhab.binding.systeminfo/2.5.0.201904152011] osgi.wiring.package; filter:="(&(osgi.wiring.package=oshi.hardware)(version>=3.13.0)(!(version>=4.0.0)))" [caused by: Unable to resolve com.github.oshi.oshi-core/3.13.0: missing requirement [com.github.oshi.oshi-core/3.13.0] osgi.wiring.package; filter:="(osgi.wiring.package=com.sun.jna.platform.linux)"]]]

Build runs smoothly. Any hint?

J-N-K commented:
Seems to be a bug in JNA 5.2.0: https://github.com/java-native-access/jna/issues/1054

Did this ever work under Linux?

J-N-K commented:
You can try removing the dependency to `net.java.dev.jna:.jna-platform`, putting ` jna-platform-5.2.0.jar` to `/lib` and add the following lines to `pom.xml`.

```
<properties>
  <bnd.exportpackage>com.sun.jna.platform.*;version="5.2.0"</bnd.exportpackage>
</properties>
```

I did not try but that should export all packages.

falkena commented:
Hi @J-N-K , thank you :-) I've traced the dependecies down to jna 4.2.1: still the same problem. It seems, that it's broken for the long time and nobody catched the bug. On Raspberry jna 5.2.0 runs as expected. Therefore it seems, that we must live with workaround.

J-N-K commented:
The export works?

falkena commented:
Build and verification was OK. Currently i rebase my other PR and test all the stuff.

falkena commented:
Ok. Rebased, updated eclipse, wait till workspace build... And new fail... 

> openhab> bundle:restart 254
> Error executing command: Error restarting bundles:
>         Unable to start bundle 254: Could not resolve module: org.openhab.binding.systeminfo [254]
>   Unresolved requirement: Import-Package: com.sun.jna; version="[5.2.0,6.0.0)"

Ok, embed jna-5.2.0.jar the same way as jna-platform-5.2.0.jar. Rebuilt, new attempt...
> openhab> bundle:restart 254
> Error executing command: Error restarting bundles:
>         Unable to start bundle 254: Could not resolve module: org.openhab.binding.systeminfo [254]
>   Unresolved requirement: Import-Package: oshi; version="[3.13.0,4.0.0)"

Ok, embed oshi-core-3.13.0.jar the same way as jna-5.2.0.jar. Rebuilt, new attempt...
> openhab> bundle:restart 254
> Error executing command: Error restarting bundles:
>         Unable to start bundle 254: Could not resolve module: org.openhab.binding.systeminfo [254]
>   Unresolved requirement: Import-Package: com.sun.management

Built on Windows and run on raspberry...

davidgraeff commented:
Did it work with the old buildsystem? If not we are not loosing anything if we merge this PR for now and keep investigating later

J-N-K commented:
This is probably due to the missing dependencies which are defined in the feature.

J-N-K commented on bundles/org.openhab.binding.systeminfo/.classpath:
```diff
@@ -0,0 +1,32 @@
+			<attribute name="test" value="true"/>
+		</attributes>
+	</classpathentry>
+	<classpathentry kind="output" path="target/classes"/>
```
add jna-platform as lib (see e.g. allplay)

falkena replied:
Removed in current version

J-N-K commented on bundles/org.openhab.binding.systeminfo/pom.xml:
```diff
@@ -0,0 +1,47 @@
+      <version>3.13.0</version>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
```
remove this dependency (always provided if embedded)

falkena replied:
Done

J-N-K commented on features/karaf/openhab-addons/src/main/feature/feature.xml:
```diff
@@ -855,7 +855,10 @@
         <feature>openhab-runtime-base</feature>
-        <bundle start-level="80">mvn:org.openhab.binding/org.openhab.binding.systeminfo/${project.version}</bundle>
+        <bundle dependency="true">mvn:com.github.oshi/oshi-core/3.13.0</bundle>
+        <bundle dependency="true">mvn:net.java.dev.jna/jna-platform/5.2.0</bundle>
```
remove feature dependency (always provided if embedded)

falkena commented:
@davidgraeff : Yes it run bevore @J-N-K I already performed requested changes yesterday, just not pushed them, since it will not run...

J-N-K commented:
As I said before: the feature resolution in your production setup will not work because it is using the old feature.xml. 

It should work if you install the missing bundles manually. Did you try that?

falkena commented:
I run two environments: production (2.5.0M1) and development mit latest snapshot. The strange thing is, i get 
> Unresolved requirement: Import-Package: oshi; version="[3.13.0,4.0.0)"
if i try to load the binding, but according to oshi-core manifest file, it's provided by oshi-core.jar...

J-N-K commented:
I raised a PR against your repo. With those fixes everything compiles and I verified that the binding is working on two systems (both Linux). The tests fail, though. Maybe you can have a look at that.

J-N-K commented:
@falkena, thanks to @matthiasblaesing JNA 5.3.0  was released very fast. I have updated my PR to your repo, please merge that, the tests now run fine, too.

falkena commented:
> 
> 
> @falkena, thanks to @matthiasblaesing JNA 5.3.0 was released very fast. I have updated my PR to your repo, please merge that, the tests now run fine, too.

Hi @J-N-K : I've merged your PR. Thank you for help! I'll rebase my another PR and test the stuff this week. If something will go wrong, i'll fix it. May be @dbwiddis can find some time and update the oshi dependency to JNA 5.3.0. Officially OSHI depends now still on JNA 5.2.0

dbwiddis commented:
Hah. I’ve been waiting on JNA 5.3 and checking almost daily. But missed a day.

J-N-K commented:
But it works with 5.3.0. so no need to hurry.
